File: backend/app/forecast.py
"""価格予測・需要予測エンジン（Phase 2、docs/development_plan.md セクション 41・88）。

過去の時系列（price_history）から最小二乗法でトレンドを推定し、季節成分を重ねて
先 N か月を予測する。AI ではなく決定論的な統計手法で確定させる（原則: セクション 93）。
予測は参考値であり、信頼度（trend の当てはまり）を併記する。
"""

from __future__ import annotations

from dataclasses import dataclass

from .history import _seasonal_component, _shift_month

# 予測する先の月数。
FORECAST_MONTHS = 6


@dataclass(frozen=True)
class ForecastPoint:
    year: int
    month: int
    value: int


@dataclass(frozen=True)
class ForecastResult:
    points: list[ForecastPoint]
    slope_per_month: float  # 1 か月あたりの変化量（円 or 指数）。
    confidence: int  # 0-100。トレンドの当てはまり（決定係数 R^2 ベース）。


def _linear_fit(values: list[float]) -> tuple[float, float, float]:
    """x=0..n-1 に対する最小二乗直線 (slope, intercept, r_squared) を返す。"""
    n = len(values)
    if n < 2:
        return 0.0, (values[0] if values else 0.0), 0.0
    xs = list(range(n))
    mean_x = sum(xs) / n
    mean_y = sum(values) / n
    sxx = sum((x - mean_x) ** 2 for x in xs)
    sxy = sum((x - mean_x) * (y - mean_y) for x, y in zip(xs, values))
    slope = sxy / sxx if sxx else 0.0
    intercept = mean_y - slope * mean_x
    ss_tot = sum((y - mean_y) ** 2 for y in values)
    ss_res = sum((y - (slope * x + intercept)) ** 2 for x, y in zip(xs, values))
    r_squared = 1 - ss_res / ss_tot if ss_tot else 0.0
    return slope, intercept, max(0.0, r_squared)


def _forecast(
    values: list[float],
    months: list[int],
    seasonality: str,
    last_year: int,
    last_month: int,
    seasonal_amplitude: float,
    horizon: int,
    lower: int,
    upper: int | None,
) -> ForecastResult:
    """時系列から先 horizon か月を予測する共通ロジック。"""
    slope, intercept, r_squared = _linear_fit(values)
    n = len(values)
    # 直近の季節成分を差し引いた基準線からトレンドを延長し、将来月の季節成分を再付与する。
    points: list[ForecastPoint] = []
    for step in range(1, horizon + 1):
        year, month = _shift_month(last_year, last_month, step)
        trend = slope * (n - 1 + step) + intercept
        seasonal = _seasonal_component(month, seasonality) * seasonal_amplitude
        value = round(trend + seasonal)
        value = max(lower, value)
        if upper is not None:
            value = min(upper, value)
        points.append(ForecastPoint(year=year, month=month, value=value))
    return ForecastResult(points=points, slope_per_month=round(slope, 2), confidence=round(r_squared * 100))


def forecast_price(
    prices: list[int],
    months_of_year: list[int],
    seasonality: str,
    last_year: int,
    last_month: int,
    horizon: int = FORECAST_MONTHS,
) -> ForecastResult:
    """価格の先読み。季節振幅は直近価格の 8%。"""
    amplitude = (prices[-1] * 0.08) if prices else 0.0
    return _forecast(
        [float(p) for p in prices], months_of_year, seasonality, last_year, last_month, amplitude, horizon, 1, None
    )


def forecast_demand(
    demand: list[int],
    months_of_year: list[int],
    seasonality: str,
    last_year: int,
    last_month: int,
    horizon: int = FORECAST_MONTHS,
) -> ForecastResult:
    """需要指数の先読み。0-100 にクランプ、季節振幅は 12 ポイント。"""
    return _forecast(
        [float(d) for d in demand], months_of_year, seasonality, last_year, last_month, 12.0, horizon, 0, 100
    )

File: src/lib/research/forecast.ts
/**
 * 価格予測・需要予測（Phase 2、docs/development_plan.md セクション 41・88）。
 *
 * backend/app/forecast.py と同一の統計手法（最小二乗トレンド + 季節成分）を移植する。
 * AI ではなく決定論的に確定し、R² ベースの信頼度を併記する（原則: セクション 93）。
 */

import { pyRound, seasonalComponent, shiftMonth } from "./history";

/** 予測する先の月数。 */
export const FORECAST_MONTHS = 6;

/** 予測時系列の内部表現。 */
export interface ForecastResultPoint {
  year: number;
  month: number;
  value: number;
}

/** 予測結果（傾き・信頼度付き）。 */
export interface ForecastResult {
  points: ForecastResultPoint[];
  slopePerMonth: number;
  confidence: number;
}

/** x=0..n-1 に対する最小二乗直線 (slope, intercept, r²) を返す。 */
export function linearFit(values: number[]): [number, number, number] {
  const n = values.length;
  if (n < 2) return [0, values[0] ?? 0, 0];
  const xs = Array.from({ length: n }, (_, i) => i);
  const meanX = xs.reduce((a, b) => a + b, 0) / n;
  const meanY = values.reduce((a, b) => a + b, 0) / n;
  const sxx = xs.reduce((acc, x) => acc + (x - meanX) ** 2, 0);
  const sxy = xs.reduce((acc, x, i) => acc + (x - meanX) * (values[i] - meanY), 0);
  const slope = sxx ? sxy / sxx : 0;
  const intercept = meanY - slope * meanX;
  const ssTot = values.reduce((acc, y) => acc + (y - meanY) ** 2, 0);
  const ssRes = values.reduce((acc, y, i) => acc + (y - (slope * i + intercept)) ** 2, 0);
  const rSquared = ssTot ? 1 - ssRes / ssTot : 0;
  return [slope, intercept, Math.max(0, rSquared)];
}

function runForecast(
  values: number[],
  seasonality: string,
  lastYear: number,
  lastMonth: number,
  seasonalAmplitude: number,
  horizon: number,
  lower: number,
  upper: number | null,
): ForecastResult {
  const [slope, intercept, rSquared] = linearFit(values);
  const n = values.length;
  const points: ForecastResultPoint[] = [];
  for (let step = 1; step <= horizon; step++) {
    const [year, month] = shiftMonth(lastYear, lastMonth, step);
    const trend = slope * (n - 1 + step) + intercept;
    const seasonal = seasonalComponent(month, seasonality) * seasonalAmplitude;
    let value = pyRound(trend + seasonal);
    value = Math.max(lower, value);
    if (upper !== null) value = Math.min(upper, value);
    points.push({ year, month, value });
  }
  return { points, slopePerMonth: Math.round(slope * 100) / 100, confidence: Math.round(rSquared * 100) };
}

/** 価格の先読み。季節振幅は直近価格の 8%。 */
export function forecastPrice(
  prices: number[],
  seasonality: string,
  lastYear: number,
  lastMonth: number,
  horizon = FORECAST_MONTHS,
): ForecastResult {
  const amplitude = prices.length ? prices[prices.length - 1] * 0.08 : 0;
  return runForecast(prices, seasonality, lastYear, lastMonth, amplitude, horizon, 1, null);
}

/** 需要指数の先読み。0-100 にクランプ、季節振幅は 12 ポイント。 */
export function forecastDemand(
  demand: number[],
  seasonality: string,
  lastYear: number,
  lastMonth: number,
  horizon = FORECAST_MONTHS,
): ForecastResult {
  return runForecast(demand, seasonality, lastYear, lastMonth, 12, horizon, 0, 100);
}
